### PR TITLE
CHANGE: @W-16263070@: Move getWorkspaceRoot to the Workspace class officially and fix edge case

### PR DESCRIPTION
**Pull request by stephen-carter-at-sf**

### CODE REVIEW

File: packages/code-analyzer-engine-api/src/workspace.ts [hunk 1]
```diff
@@ -81,4 +122,44 @@ export async function expandToListAllFiles(absoluteFileOrFolderPaths: string[]):
...
     return allFiles.sort();
+}
+
+/**
+ * Returns the longest common parent folder of the provided paths.
+ *   If empty or if no common parent folder exists, like in the case on Windows machines of using two different drives
+ *   C: and D:, then null is returned (to allow the client of this function to handle this case without try/catch).
+ * @param paths It is assumed that paths is a non-empty array of absolute value paths.
+ */
+export function calculateLongestCommonParentFolderOf(paths: string[]): string | null {
```
Comment (stephen-carter-at-sf) [inline packages/code-analyzer-engine-api/src/workspace.ts:133]: Note that this all basically just moved from the eslint utils. Mostly unchanged.

File: packages/code-analyzer-engine-api/src/workspace.ts [hunk 2]
```diff
@@ -81,4 +122,44 @@ export async function expandToListAllFiles(absoluteFileOrFolderPaths: string[]):
...
+        return null;
+    }
+    const longestCommonStr: string = getLongestCommonPrefix(paths);
+    if (longestCommonStr.length === 0) {
+        return null;
+    }
+    if (longestCommonStr.length > 1 && longestCommonStr.endsWith(path.sep)) {
+        return longestCommonStr.slice(0, longestCommonStr.length - 1);
+    }
+    return fs.existsSync(longestCommonStr) && fs.statSync(longestCommonStr).isDirectory() && !isPartialFileName(longestCommonStr, paths) ?
```
Comment (stephen-carter-at-sf) [inline packages/code-analyzer-engine-api/src/workspace.ts]: The !isPartialFileName(longestCommonStr, paths) is all that was added to this. It helps handle a unique edge case which I locked in with a bunch more tests. Otherwise, everything else is unchanged.

File: packages/code-analyzer-engine-api/test/workspace.test.ts [hunk 3]
```diff
@@ -39,9 +45,112 @@ describe('Tests for the Workspace class', () => {
...
+        ]);
+        expect(workspace2.getWorkspaceRoot()).toEqual(path.join(SAMPLE_WORKSPACE_FOLDER, 'sub1', 'sub2'));
+    });
+
+    it('When root folder of workspace is also partially the name of another file that is not in the workspace, then getWorkspaceRoot returns the folder', () => {
+        const workspace: Workspace = new Workspace([
+            path.join(SAMPLE_WORKSPACE_FOLDER, 'sub1', 'sub2', 'someFile') // This is a folder. Notice the someFile1InSub2.txt is not our workspace
+        ]);
+        expect(workspace.getWorkspaceRoot()).toEqual(path.join(SAMPLE_WORKSPACE_FOLDER, 'sub1', 'sub2', 'someFile'));
+    });
```
Comment (stephen-carter-at-sf) [inline packages/code-analyzer-engine-api/test/workspace.test.ts]: These are the new tests that handle the edge cases that weren't handled before.

File: packages/code-analyzer-eslint-engine/test/utils.test.ts [hunk 4]
```diff
@@ -1,57 +1,4 @@
-import {calculateLongestCommonParentFolderOf, makeUnique} from "../src/utils";
-import path from "node:path";
-
-describe('Tests for the calculateLongestCommonParentFolderOf utility function', () => {
```
Comment (stephen-carter-at-sf) [inline packages/code-analyzer-eslint-engine/test/utils.test.ts:4]: All these tests moved to the engine api's workspace.test.ts file essentially.

File: packages/code-analyzer-engine-api/src/workspace.ts [hunk 5]
```diff
@@ -81,4 +122,44 @@ export async function expandToListAllFiles(absoluteFileOrFolderPaths: string[]):
...
+    // Then we check that each string's ith character is the same as the shortest strings ith character
+    for (let i = 0; i < shortestStr.length; i++) {
+        if(!strs.every(str => str[i] === shortestStr[i])) {
+            // If we find a string that doesn't match the ith character, we return the common prefix from [0,i)
+            return shortestStr.substring(0, i)
+        }
+    }
+    return shortestStr;
+}
+function isPartialFileName(fileOrFolder:string, paths: string[]) {
```
Comment (jag-j) [inline packages/code-analyzer-engine-api/src/workspace.ts]: Nit - Can this function name be a bit more descriptive? I had to read the comments and the tests to figure out what it is doing.

**Review by jag-j — APPROVED** [review]

**Review by jag-j — APPROVED** [review]

**Review by jag-j — APPROVED** [review]

File: packages/code-analyzer-engine-api/src/workspace.ts (hunk 5 above)
Comment (stephen-carter-at-sf) [reply packages/code-analyzer-engine-api/src/workspace.ts]: Sure... renamed this to includesAFileThatIsNotAFolderThatStartsWith

**Review by jfeingold35 — APPROVED** [review]
